inc/namespace.php

```diff
@@ -63,3 +87,40 @@ function basic_auth_sanitization_callback( $value ) : string {
 
 	return 'on';
 }
+
+/**
+ * Require PHP Basic authentication.
+ *
+ * Basic auth username and password should be set wherever constants for the site are defined.
+ */
+function require_auth() {
+	$basic_auth = get_option( 'hm-basic-auth' );
+
+	if (
+		// Bail if basic auth has been disabled...
+		( ! $basic_auth || 'off' === $basic_auth ) ||
+		// ...or if the development environment isn't defined or explicitly false.
+		( ! is_development_environment() )
+	) {
+		return;
+	}
+
+	// Check for a basic auth user and password.
+	if ( defined( 'HM_BASIC_AUTH_PW' ) && defined( 'HM_BASIC_AUTH_USER' ) ) {
+		header( 'Cache-Control: no-cache, must-revalidate, max-age=0' );
+		$site_name = get_option( 'blogname' );
+
+		$has_supplied_credentials = ! empty( $_SERVER['PHP_AUTH_USER'] ) && ! empty( $_SERVER['PHP_AUTH_PW'] );
+		$is_not_authenticated     = (
+			! $has_supplied_credentials ||
+			$_SERVER['PHP_AUTH_USER'] !== HM_BASIC_AUTH_USER ||
+			$_SERVER['PHP_AUTH_PW'] !== HM_BASIC_AUTH_PW
+		);
+
+		if ( $is_not_authenticated ) {
+			header( 'HTTP/1.1 401 Authorization Required' );
+			header( "WWW-Authenticate: Basic realm=\"$site_name development site login\"" );
```

Comment:
I'll create new issues for these. At least some of these could probably sneak in sooner than later (like i18n and escaping) since we may need them for SIEJ.